### PR TITLE
ldap sync improvements

### PR DESCRIPTION
**Pull request by deads2k**
This addresses a few comments made by @liggitt.  Most importantly, it provides feedback while the job is running, since this likely to be long running.

It also adds more unit tests, eliminates unnecessary cruft from options objects, makes the create/update flow more straightforward, and fixes a testclient problem that @kargakis found (let me know if you want that separately).

@stevekuznetsov @pweil- ptal.

### CODE REVIEW

**Comment by deads2k** [comment]
I'll fix the double error printing tomorrow.

[test][extended:authentication]

File: pkg/cmd/experimental/syncgroups/groupsyncer.go [hunk 1]
```diff
@@ -141,39 +134,31 @@ func (s *LDAPGroupSyncer) findOpenShiftGroup(ldapGroupUID string) (*userapi.Grou
...
+	// make sure we aren't taking over an OpenShift group that is already related to a different LDAP group
+	if url, exists := group.Annotations[ldaputil.LDAPURLAnnotation]; !exists || (url != s.Host) {
+		return nil, fmt.Errorf("group %q: %s annotation did not match sync host: wanted %s, got %s",
 			group.Name, ldaputil.LDAPURLAnnotation, s.Host, url)
 	}
-	uid, exists := group.Annotations[ldaputil.LDAPUIDAnnotation]
-	if !exists || uid != ldapGroupUID {
-		return nil, fmt.Errorf("group %s's %s annotation did not match LDAP UID: wanted %s, got %s",
+	if uid, exists := group.Annotations[ldaputil.LDAPUIDAnnotation]; !exists || (uid != ldapGroupUID) {
+		return nil, fmt.Errorf("group %q: %s annotation did not match LDAP UID: wanted %s, got %s",
```
Comment (stevekuznetsov) [inline pkg/cmd/experimental/syncgroups/groupsyncer.go]: Do we need to change anything else here to support that migrating LDAP group case you are worried about?

File: pkg/cmd/experimental/syncgroups/groupsyncer_test.go [hunk 2]
```diff
@@ -1,22 +1,115 @@
...
+
+		actualGroup, err := syncer.makeOpenShiftGroup(tc.ldapGroupUID, tc.usernames)
+		if err != nil && len(tc.expectedErr) == 0 {
+			t.Errorf("%s: unexpected error %v", name, err)
+
+		} else if err == nil && len(tc.expectedErr) != 0 {
+			t.Errorf("%s: expected %v, got nil", name, tc.expectedErr)
+
+		} else if err != nil {
+			if e, a := tc.expectedErr, err.Error(); e != a {
```
Comment (stevekuznetsov) [inline pkg/cmd/experimental/syncgroups/groupsyncer_test.go]: expected, actual, testCase? Naming here makes it hard to follow for what that's worth.

File: pkg/cmd/experimental/syncgroups/groupsyncer_test.go [hunk 3]
```diff
@@ -77,54 +170,94 @@ var Group1Members []*ldap.Entry = []*ldap.Entry{Member1, Member2}
...
-		GroupMemberExtractor: &testGroupMemberExtractor,
-		UserNameMapper:       &testUserNameMapper,
-		GroupNameMapper:      &testGroupNameMapper,
-		GroupClient:          &testGroupClient,
-		Host:                 testHost,
-		SyncExisting:         false,
+func groupExists(haystack []*userapi.Group, needle *userapi.Group) bool {
+	for _, actual := range haystack {
+		t := *actual
+		actualGroup := &t
```
Comment (stevekuznetsov) [inline pkg/cmd/experimental/syncgroups/groupsyncer_test.go]: What is this pointer referencing-dereferencing doing? Can you rename variables here to make it more obvious?

File: pkg/cmd/experimental/syncgroups/groupsyncer_test.go [hunk 4]
```diff
@@ -77,54 +170,94 @@ var Group1Members []*ldap.Entry = []*ldap.Entry{Member1, Member2}
...
-		NameMapping: map[string]string{
-			Group1UID: "os" + Group1UID,
-			Group2UID: "os" + Group2UID,
-			Group3UID: "os" + Group3UID,
-		},
+}
+
+func TestMissingLDAPGroupUIDMapping(t *testing.T) {
+	testGroupSyncer, tc := newTestSyncer()
+	testGroupSyncer.GroupLister.(*TestGroupLister).GroupUIDs = append(testGroupSyncer.GroupLister.(*TestGroupLister).GroupUIDs, "ldapgroupwithnouid")
```
Comment (stevekuznetsov) [inline pkg/cmd/experimental/syncgroups/groupsyncer_test.go]: Can you throw this into a closure around `newTestSyncer()` instead of this nonsense?

File: pkg/cmd/experimental/syncgroups/groupsyncer_test.go [hunk 5]
```diff
@@ -77,54 +170,94 @@ var Group1Members []*ldap.Entry = []*ldap.Entry{Member1, Member2}
...
-		MemberMapping: map[string][]*ldap.Entry{
-			Group1UID: Group1Members,
-			Group2UID: Group2Members,
-			Group3UID: Group3Members,
-		},
+
+	checkClientForGroups(tc, newDefaultOpenShiftGroups(testGroupSyncer.Host), t)
+}
+
+func TestListFails(t *testing.T) {
```
Comment (stevekuznetsov) [inline pkg/cmd/experimental/syncgroups/groupsyncer_test.go]: Why is this a valuable test?

File: pkg/cmd/experimental/syncgroups/syncgroups.go [hunk 6]
```diff
@@ -258,21 +227,19 @@ func (o *SyncGroupsOptions) Validate() error {
 func (o *SyncGroupsOptions) Run(cmd *cobra.Command, f *clientcmd.Factory) error {
 	// In order to create the GroupSyncer, we need to build its' parts:
 	// interpret user-provided configuration
-	clientConfig, err := ldaputil.NewLDAPClientConfig(
-		o.Config.URL,
-		o.Config.BindDN,
-		o.Config.BindPassword,
-		o.Config.CA,
-		o.Config.Insecure)
+	clientConfig, err := ldaputil.NewLDAPClientConfig(o.Config.URL, o.Config.BindDN, o.Config.BindPassword, o.Config.CA, o.Config.Insecure)
```
Comment (stevekuznetsov) [inline pkg/cmd/experimental/syncgroups/syncgroups.go]: We should establish some style rules that we can all follow ...

File: pkg/cmd/experimental/syncgroups/syncgroups.go [hunk 7]
```diff
@@ -88,25 +75,12 @@ type SyncGroupsOptions struct {
...
-
 	// WhitelistContents are the contents of the whitelist: names of OpenShift group or LDAP group UIDs
 	WhitelistContents []string
 
-	// SyncExisting determines that only groups in OpenShift previously synced with this LDAP server
-	// will be synced again
-	SyncExisting bool
-
 	// Confirm determines whether not to write to openshift
 	Confirm bool
```
Comment (stevekuznetsov) [inline pkg/cmd/experimental/syncgroups/syncgroups.go]: Dry-run?

File: pkg/cmd/experimental/syncgroups/groupsyncer.go [hunk 8]
```diff
@@ -88,10 +80,20 @@ func (s *LDAPGroupSyncer) GetResultingGroups() ([]*userapi.Group, []error) {
...
 		openshiftGroup, err := s.makeOpenShiftGroup(ldapGroupUID, usernames)
 		if err != nil {
+			fmt.Fprintf(s.Err, "Error building OpenShift group for LDAP group %q: %v.  Continuing with the next LDAP group.", ldapGroupUID, err)
 			errors = append(errors, err)
 			continue
 		}
 		openshiftGroups = append(openshiftGroups, openshiftGroup)
+
+		if !s.DryRun {
+			fmt.Fprintf(s.Out, "group/%s\n", openshiftGroup.Name)
```
Comment (stevekuznetsov) [inline pkg/cmd/experimental/syncgroups/groupsyncer.go]: Can we do this with the printer instead of in here?

**Comment by stevekuznetsov** [comment]
Bash completions out of date

File: pkg/cmd/experimental/syncgroups/syncgroups.go [hunk 9]
```diff
@@ -158,9 +136,9 @@ func NewCmdSyncGroups(name, fullName string, f *clientcmd.Factory, out io.Writer
 		},
 	}
 
-	cmd.Flags().StringVar(&options.WhitelistSource, "whitelist", "", "path to the group whitelist")
-	cmd.Flags().StringVar(&options.ConfigSource, "sync-config", "", "path to the sync config")
-	cmd.Flags().BoolVar(&options.SyncExisting, "existing", false, "sync only existing, previously-synced groups")
+	cmd.Flags().StringVar(&whitelistFile, "whitelist", whitelistFile, "path to the group whitelist")
+	cmd.Flags().StringVar(&configFile, "sync-config", configFile, "path to the sync config")
```
Comment (stevekuznetsov) [inline pkg/cmd/experimental/syncgroups/syncgroups.go]: +1 for these changes

File: pkg/cmd/experimental/syncgroups/groupsyncer.go (hunk 8 above)
Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/groupsyncer.go]: > Can we do this with the printer instead of in here?

No.  Imagine a very slow LDAP server.  This sync job might take 20 minutes.  In that case, we want to assure the user that something is happening, so they won't cancel it.  This will be ticking off the groups as we go through them.  We use this format so that it can be piped to other commands.

File: pkg/cmd/experimental/syncgroups/syncgroups.go (hunk 7 above)
Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/syncgroups.go]: > Dry-run?

I debated, but decided to match the flag for ease of binding.  We've previously decided that commands that default to `--dry-run=true`, should have a `--confirm` flag instead (it allows `oc --confirm` instead of `oc --dry-run=false`.).  `Confirm` looked stupid in the syncer, but I'm open to other names in the syncer if you want to avoid the later `!`.

File: pkg/cmd/experimental/syncgroups/groupsyncer_test.go (hunk 5 above)
Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/groupsyncer_test.go]: > Why is this a valuable test?

It makes sure we exit in a controlled way as expected and don't progress the job any further on failed lists.

File: pkg/cmd/experimental/syncgroups/groupsyncer_test.go (hunk 3 above)
Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/groupsyncer_test.go]: > What is this pointer referencing-dereferencing doing? Can you rename variables here to make it more obvious?

It makes a copy of the data and then points back to it so that you can change the copy without changing the original.

This one in particular is bugged because the map it manipulates is actually a pointer which is pulled into the shallow copy.   Updated.

File: pkg/cmd/experimental/syncgroups/groupsyncer_test.go (hunk 2 above)
Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/groupsyncer_test.go]: > expected, actual, testCase? Naming here makes it hard to follow for what that's worth.

Those are pretty common in our codebase.  I blame liggitt and ironcladlou

File: pkg/cmd/experimental/syncgroups/groupsyncer.go (hunk 1 above)
Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/groupsyncer.go]: > Do we need to change anything else here to support that migrating LDAP group case you are worried about?

Maybe, but I'll keep that change separate.  I think it might be more easily done with a scripted `patch`.

File: pkg/cmd/experimental/syncgroups/groupsyncer_test.go (hunk 2 above)
Comment (liggitt) [inline pkg/cmd/experimental/syncgroups/groupsyncer_test.go]: I've never intentionally done that

**Comment by deads2k** [comment]
re[test]

File: pkg/cmd/experimental/syncgroups/syncgroups.go [hunk 10]
```diff
@@ -376,7 +332,7 @@ func (o *SyncGroupsOptions) Run(cmd *cobra.Command, f *clientcmd.Factory) error
 		return err
```
Comment (stevekuznetsov) [inline pkg/cmd/experimental/syncgroups/syncgroups.go]: How is this print here going to conflict with the print during the sync job?

How do we ensure that someone trying to pipe the output of this sync command to another doesn't have to deal with other garbage?

Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/syncgroups.go]: > How is this print here going to conflict with the print during the sync job?
> 
> How do we ensure that someone trying to pipe the output of this sync command to another doesn't have to deal with other garbage?

They look mutually exclusive to me.  We'll exit in the `if` above if we got the printed output.

**Comment by deads2k** [comment]
data race in kube.  re[test]

**Comment by stevekuznetsov** [comment]
All small comments on variable names and help text we can fix later, otherwise LGTM

**Comment by deads2k** [comment]
[merge]
